tests/e2e/k8s/suite.go

```diff
@@ -0,0 +1,190 @@
+	Port:      80,
+}
+
+var httpEchoService = util.Service{
+	Name:      "http-echo",
+	Namespace: v1.NamespaceDefault,
+	Port:      80,
+}
+
+func httpEchoPodAndService(echoValue string) *util.PodAndService {
+	return &util.PodAndService{
+		Service:    httpEchoService,
+		Image:      "hashicorp/http-echo",
+		Args:       []string{"-listen=:8080", "-text=" + echoValue},
+		TargetPort: 8080,
+	}
+}
+
+// TestSuite is a suite for e2e testing on k8s clusters.
+type TestSuite struct {
+	suite.Suite
+	fabric   *util.Fabric
+	clusters []*util.KindCluster
+}
+
+// SetupSuite creates the k8s kind clusters and the clusterlink certificates.
+func (s *TestSuite) SetupSuite() {
+	fabric, err := util.NewFabric()
+	if err != nil {
+		s.T().Fatal(err)
+	}
+
+	// create clusters and fabric
+	s.fabric = fabric
+	s.clusters = make([]*util.KindCluster, clusterCount)
+	for i := 0; i < clusterCount; i++ {
+		s.clusters[i] = util.NewKindCluster(fmt.Sprintf("peer%d", i))
+		fabric.CreatePeer(s.clusters[i])
+
+		s.clusters[i].Start()
+		for _, image := range images {
+			s.clusters[i].LoadImage(image)
+		}
+	}
+
+	// prepare logs directory
+	if err := os.RemoveAll(exportedLogsPath); err != nil {
+		s.T().Fatal(fmt.Errorf("cannot cleanup logs directory: %w", err))
+	}
+	if err := os.MkdirAll(exportedLogsPath, 0755); err != nil {
+		s.T().Fatal(fmt.Errorf("cannot create logs directory: %w", err))
+	}
+
+	// wait for clusters
+	for i := 0; i < clusterCount; i++ {
+		if err := s.clusters[i].Wait(); err != nil {
+			s.T().Fatal(err)
+		}
+
+		// create http-echo service which echos the cluster name
+		err := s.clusters[i].CreatePodAndService(httpEchoPodAndService(s.clusters[i].Name()))
+		if err != nil {
+			s.T().Fatal(fmt.Errorf("cannot create http-echo service: %w", err))
+		}
+
+		// create iperf3 server service
+		err = s.clusters[i].CreatePodAndService(&util.PodAndService{
+			Service:    iperfService,
+			Image:      "networkstatic/iperf3",
+			Args:       []string{"-s", "-p", "8080"},
+			TargetPort: 8080,
+		})
+		if err != nil {
+			s.T().Fatal(fmt.Errorf("cannot create iperf3-server service: %w", err))
+		}
+	}
+
+	// wait for fabric
+	if err := fabric.Wait(); err != nil {
+		s.T().Fatal(err)
+	}
+}
+
+// TearDownSuite deletes the k8s kind clusters.
+func (s *TestSuite) TearDownSuite() {
+	for _, cluster := range s.clusters {
+		if err := cluster.Destroy(); err != nil {
+			s.T().Fatal(err)
+		}
+	}
+}
+
+// convert e.g. TestBlaBla to test-bla-bla
+func convertCaseCamelToKebab(s string) string {
+	s = regexp.MustCompile("(.)([A-Z][a-z]+)").ReplaceAllString(s, "${1}-${2}")
+	s = regexp.MustCompile("([a-z0-9])([A-Z])").ReplaceAllString(s, "${1}-${2}")
+	return strings.ToLower(s)
+}
+
+// BeforeTest creates the test namespace before each test.
+func (s *TestSuite) BeforeTest(_, testName string) {
+	testName = convertCaseCamelToKebab(testName)
+	if err := s.fabric.SwitchToNewNamespace(testName, false); err != nil {
+		s.T().Fatal(err)
+	}
```

Comment:
default service account is created when the kind clusters are created, and as I said in other comment, I don't think it's necessary to wait for it.

Reply:
I don't think this is true.
The `default` service account is created per namespace (i.e., ns1/default, ns2/default...). There a k8s controller that watches for namespaces and creates the default service account in each namespace. This is done async by the controller. They are not created on cluster startup.
We had CICD failures when Pod creation was "too fast" and the controller did not finish creating the SA. This seems to be a relatively [common failure](https://www.google.com/search?q=kind+serviceaccount+default+not+found) in tests suites, so might be worthwhile to check

Reply:
I think the controlplane should gracefully handle a case where the default service account is not yet created.
Assuming it does, are there any other pitfalls?

Reply:
which control plane - clusterlink or k8s?
I'm not aware of additional pitfalls *besides Pods failing to start), but the assumption is not correct.

Dealing with creation of SAs is the responsibility of a built-in k8s controller. It is not something the clusterlink CP deal with. Non-default SAs are the user's responsibility.
The k8s API does not deal with it gracefully and will reject/fail Pod creations that use a non-existent SA.
The way to "deal" with it is (1) pray (with or without a "sleep") we don't see the race condition (I think this will be the case 99.9% of the time - no noticeable failure); (2) wait for the existence of the default service account in the current namespace before using it in Pod definitions. 
I'm ok walking into this potential race (i.e., do nothing now) and monitoring CICD for this type of failure.

Reply:
Added a check for verifying default service account exist before deploying clusterlink.
Thanks!